glib/src/match_info.rs

```diff
@@ -1,9 +1,380 @@
 // Take a look at the license at the top of the repository in the LICENSE file.
 
-use crate::{translate::*, IntoGStr, MatchInfo};
+use crate::{translate::*, GStr, IntoGStr, Regex};
 use std::{mem, ptr};
 
-impl MatchInfo {
+#[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+#[repr(transparent)]
+pub struct MatchInfo<'input> {
+    inner: crate::shared::Shared<ffi::GMatchInfo, Self>,
```

Comment:
```suggestion
    inner: std::ptr::NonNull<ffi::GMatchInfo>,
```

should be sufficient or not? Then the `SharedMemoryManager` noise below can also go away

Reply:
That would work too, but would require manually implementing the reference-counting logic that `Shared` already implements, i.e. calling `g_match_info_ref` in `clone` instead of deriving `Clone`, and implementing `Drop` to call `g_match_info_unref`.

Reply:
I guess the question then is which is simpler. Your call

Reply:
I went with using `NonNull` and manually implementing the reference counting, since `Shared` also caused issues with variance (`Shared<_, MM>` is invariant in `MM`, meaning holding a `Shared<_, Self>` in `MatchInfo<'input>` causes `MatchInfo` to be invariant in `'input`, when ideally it should be covariant).